File: Problem Solving/boj/random defense/1958-05-17-2023.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static String[] str = new String[3];
    static HashMap<String, Integer> map = new HashMap<>();

    public static void main(String[] args) throws IOException {
        for (int i = 0; i < 3; i++)
            str[i] = br.readLine();

        Integer ans = 0;
        for (int i = 0; i < 3; i++) {
            int code = (1 << i);
            for (int len = 1; len <= str[i].length(); len++) {
                for (int stidx = 0; stidx + len <= str[i].length(); stidx++) {
                    String substr = str[i].substring(stidx, stidx + len);

                    Integer foundValue = map.get(substr);
                    if (foundValue == null) {
                        map.put(substr, code);
                        foundValue = code;
                    }
                    int valueToBeUpdated = (foundValue | code);
                    if (valueToBeUpdated == 7) {
                        ans = Math.max(ans, len);
                        System.out.println(substr);
                    }
                    map.put(substr, valueToBeUpdated);
                }
            }
        }
        System.out.println(ans);
    }
}

File: Problem Solving/boj/random defense/1958-05-17-2023.cpp
#include <bits/stdc++.h>
using namespace std;

unordered_map<string, int> um;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  string s; int ans = 0;
  for (int i = 0; i < 3; i++) {
    cin >> s;
    int code = (1 << i);
    for (int len = 1; len <= s.length(); len++) {
      for (int st = 0; st + len <= s.length(); st++) {
        string ss = s.substr(st, len);
        um[ss] = (um[ss] | code);
        if (um[ss] == 7) ans = max(ans, len);
      }
    }
  }
  cout << ans;
}